tsdb/shard.go

```diff
@@ -2040,154 +2086,247 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
+			err = fmt.Errorf("failed removing saved field changes - %s: %w", fs.changeMgr.changeFilePath, e)
 		}
 	}()
+	fs.mu.RLock()
+	defer fs.mu.RUnlock()
 	isEmpty, err := func() (isEmpty bool, err error) {
 		// ensure temp file closed before rename (for Windows)
 		defer func() {
-			if e := fd.Close(); err == nil {
-				err = e
+			if e := fd.Close(); err == nil && e != nil {
+				err = fmt.Errorf("closing %s: %w", path, e)
 			}
 		}()
 		if _, err = fd.Write(fieldsIndexMagicNumber); err != nil {
-			return true, err
+			return true, fmt.Errorf("failed writing magic number for %s: %w", path, err)
 		}
 
-		// Read all the pending new field and measurement write requests
-		// that will be captured in the marshaling of the in-memory copy
-		for {
-			select {
-			case ec := <-fs.writer.writeRequests:
-				errorChannels = append(errorChannels, ec.done)
-				continue
-			default:
-			}
-			break
-		}
 		// Lock, copy, and marshal the in-memory index
-		b, err := fs.marshalMeasurementFieldSet()
+		b, err := fs.marshalMeasurementFieldSetNoLock()
 		if err != nil {
-			return true, err
+			return true, fmt.Errorf("failed marshaling fields for %s: %w", fs.path, err)
 		}
 		if b == nil {
 			// No fields, file removed, all done
 			return true, nil
 		}
 		if _, err := fd.Write(b); err != nil {
-			return true, err
+			return true, fmt.Errorf("failed saving fields to %s: %w", path, err)
 		}
-		return false, fd.Sync()
+		return false, nil
 	}()
-	if err != nil || isEmpty {
-		return
+	if err != nil {
+		return err
+	} else if isEmpty {
+		// remove empty file
+		if err = os.RemoveAll(fs.path); err != nil {
+			return fmt.Errorf("cannot remove %s: %w", fs.path, err)
+		} else {
+			return nil
+		}
 	}
-	err = fs.renameFile(path)
+
+	return fs.renameFileNoLock(path)
 }
 
-// marshalMeasurementFieldSet: remove the fields.idx file if no fields
-// otherwise, copy the in-memory version into a protobuf to write to
-// disk
-func (fs *MeasurementFieldSet) marshalMeasurementFieldSet() ([]byte, error) {
-	fs.mu.Lock()
-	defer fs.mu.Unlock()
-	if len(fs.fields) == 0 {
-		// If no fields left, remove the fields index file
-		if err := os.RemoveAll(fs.path); err != nil {
-			return nil, err
+// appendToChangesFile: Write a change file for fields.idx
+// Only called in one Go proc, so does not need locking.
+func (fscm *measurementFieldSetChangeMgr) appendToChangesFile(first writeRequest) {
+	var err error = nil
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels := []errorChannel{first.errorReturn}
+	changes := []FieldChanges{first.changes}
+	// On return, send the error to every go proc that send changes
+	defer func() {
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
+	}()
+	log, end := logger.NewOperation(context.TODO(), fscm.logger, "saving field index changes", "MeasurementFieldSet")
```

Comment:
Are you intentionally leaving this TODO here in case this method eventually takes a context?

Reply:
I didn't want to pull a context down from some caller above, because I didn't want need cancellation or such, but let me look at if there is a an easy way to get one into there for future use...  Good question, thanks!

Reply:
@jeffreyssmith2nd - There's no easy context to pull into where I had the `TODO()` calls.  They are just used to store a trace ID for the operation, so I could make then into `Background` calls or leave them as `TODO`.  What do you prefer?

Reply:
I think a `TODO` is probably fine here, since it's really just being used to satisfy an API need and the difference between `Background` and `TODO` is entirely semantics anyways.